default/python/universe_generation/universe_generator.py

```diff
@@ -144,6 +144,19 @@ def create_universe(psd_map):
     seed_rng(seed_pool.pop())
     distribute_specials(gsd.specials_frequency, fo.get_all_objects())
 
+    # set game uid
+    empire_names = []
+    for psd in psd_map.values():
+        empire_names.append(psd.empire_name)
+
+    empire_names.sort()
+    for i, v in enumerate(empire_names):
+        empire_names[i] = v.capitalize()[:2]
+
+    fo.get_galaxy_setup_data().gameUID = "".join(empire_names) + str(random.randint(0,999)).zfill(3)
```

Comment:
You use pseudo-random number generator that initialized outside this code. It may produce always the same number for same universe seed (now or after some changes). It looks dangerous to me.

Maybe you should create own generator and initialize it with seed that uses current timestamp. IMHO names + time with milliseconds will be unique enough.

Reply:
OR just use https://docs.python.org/2/library/uuid.html#uuid.uuid3

Reply:
UUID is hard to memorize. But thanks for info about that `random`. I'll try to change it to timestamp seeded random.

Reply:
I don't think the universe generation Python scripts are the right place to generate the UUID, that should be done in the C++ backend code.

The universe generation scripts are, like the FOCS scripts, user customizable content, and should only contain stuff that needs to be user customizable. IMO that's not the case with the UUID, so no reason to rely on user customizable scripts to properly initialize it.

What's done in the C++ code can't be broken by users tinkering with the Pytho scripts.

Reply:
That would also solve the problem of using an independent PRNG. If you reseed the Python PRNG with a timestamp derived seed here, that can/will mess up reproducible universe generation.

Reply:
I'm going to get UID from DBMS to ensure uniqueness and use it as a primary and foreign key when store games' statistics.
So I prefer to don't move DBMS support into C++ backend code.

Reply:
Why do you want to get the UID from a DBMS? Uniqueness can be ensured by a properly implemented C++ function too, that's not something which can only be done via a DBMS.

That said, if you absolutely want to, you can still generate a UID in the universe generation Python code, and overwrite an UID that has been generated in the C++ code. But I would still have a UID generated in the C++ code, so a game gets a proper UID even if the universe generation scripts fail to do generate one.

Unless the UID only gets used in the Python scripts, and is never used/referenced in the C++ backend code. In that case generating the UID only in the Python scripts is fine of course.

Reply:
In C++ it used to determine subfolder for save game in hostless mode and also shows in galaxy setup encyclopedia for the players.

Reply:
In that case the UID really needs to be generated in the C++ code, and should not rely on user customizable scripts to do that correctly.

We can make it so that the C++ generated UID can be overwritten in the Python scripts if so desired. Of course we have to make sure the overwritten value is written back to the C++ data structures.

Reply:
It already has getter and setter Python API so the only left is the C++ initialization.

Reply:
There is one for loading old saves: https://github.com/freeorion/freeorion/blob/master/util/SerializeMultiplayerCommon.cpp#L40

Reply:
Provided PR moving game UID generation to C++: #2269.

Reply:
(@o01eg, please review)